### PR TITLE
Feature/135 tab component

### PR DESCRIPTION
**Pull request by patric-eberle**
## Pull request
Adds a new, basic tab component.

I decided to create it as an `element` component, since it most likely will be need an additional component to switch to an alternative layout on mobile.

I decided to create a Storybook demo for this component to build up my knowledge in this part of the application too.
 
### Ticket
https://github.com/valantic/vue-template/issues/135
 
### Browser testing
- http://localhost:6006/?path=/story/components-c-tabs--default
 
### Checklist
- [x] I merged the current development branch (before testing)
- [ ] Added JSDoc and styleguide demo
- [x] Tested all links in project relevant browsers
- [x] Tested all links in different screen sizes
- [x] Did run automated tests and linters
- [x] Did assign ticket
- [x] Double checked target branch
 
## Review/Test checklist
- [x] Did review code and documentation
- [x] Tested all links in project relevant browsers
- [x] Tested all links in different screen sizes
- [x] Did check accessibility (Wave, only errors)
- [ ] Re-assign ticket to developer

### CODE REVIEW

**Review by claudio147 — APPROVED** [review]

**Comment by claudio147** [comment]
@patric-eberle Please check the merge conflicts. Your changes are fine 👍

**Comment by janwidmer** [comment]
@patric-eberle regarding the decision to build it as element.. When we do that and than pass components into the tab component via slots, we have the structure:

- Component Tab Wrapper
    - Element Tab
        - Component / Element XYZ

Therfore, we have the kinda weird nesting.. Wouldn't it be better to define tabs also as component? Then we would have:

- Component Tab Wrapper
    - Component Tab
        - Component / Element XYZ

**Comment by patric-eberle** [comment]
@MB-web can you please check my latest changes. THX.

File: src/components/c-tabs.vue [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+           :class="b('tab', { [tab.id]: true, active: tab.id === activeTab.id })"
+           :aria-selected="tab.id === activeTab.id"
+           :aria-controls="`c-tabs-${uuid}--panel-${tab.id}`"
+           :href="tab.link?.href || `#${tab.id}`"
+           :target="tab.link?.target || LINK_TARGET.SELF"
+           :rel="tab.link?.target === LINK_TARGET.BLANK && 'noopener noreferrer'"
+           role="tab"
+           @click="onTabClick(tab)"
+        >
+          <!-- @slot ${tab.id}__tab - Allows to overwrite the tab content. -->
```
Comment (MB-web) [inline src/components/c-tabs.vue:21]:
```suggestion
          <!-- @slot ${tab.id}__tab - Allows to overwrite the tab title. -->
```

File: src/components/c-tabs.vue [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+  import { ILink } from '@/types/link';
+
+  interface ISetup extends IUuid {}
+
+  export interface ITab {
+
+    /**
+     * The text title for the current tab.
+     */
+    title: string,
```
Comment (MB-web) [inline src/components/c-tabs.vue:58]:
```suggestion
    title: string;
```

File: src/components/c-tabs.vue [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+    /**
+     * A unique id for the current tab.
+     */
+    id: string,
```
Comment (MB-web) [inline src/components/c-tabs.vue:63]:
```suggestion
    id: string;
```

File: src/components/c-tabs.vue [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+    /**
+     * Determines if the given tab should be initially active.
+     */
+    active?: boolean,
```
Comment (MB-web) [inline src/components/c-tabs.vue:68]:
```suggestion
    active?: boolean;
```

File: src/components/c-tabs.vue [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+    /**
+     * Allows to track additional information for a tab. e.g. a link.
+     */
+    link?: ILink,
```
Comment (MB-web) [inline src/components/c-tabs.vue:73]:
```suggestion
    link?: ILink;
```

File: src/components/c-tabs.vue [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+  }
+
+  interface IData {
+    LINK_TARGET: typeof LINK_TARGET,
+
+    /**
+     * Holds the currently active tab defintion.
+     */
+    activeTab: ITab|null,
```
Comment (MB-web) [inline src/components/c-tabs.vue:82]:
```suggestion
    activeTab: ITab | null,
```

File: src/components/c-tabs.vue [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+      tabs: {
+        type: Array as PropType<ITab[]>,
+        required: true,
+      },
+
+      /**
+       * Accepts a label that describes the purpose of the set of tabs.
+       */
+      ariaLabel: {
+        type: String,
```
Comment (MB-web) [inline src/components/c-tabs.vue:110]:
```suggestion
        type: String | null,
```

File: src/components/c-tabs.vue [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+        default: null,
+      },
+    },
+    emits: {
+      change(payload: ITab) {
```
Comment (MB-web) [inline src/components/c-tabs.vue:115]:
```suggestion
      change(payload: ITab): boolean {
```

File: src/components/c-tabs.vue [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+    // activated() {},
+    // deactivated() {},
+    // beforeUnmount() {},
+    // unmounted() {},
+
+    methods: {
+      /**
+       * Handles the click event of tabs.
+       */
+      onTabClick(tab: ITab) {
```
Comment (MB-web) [inline src/components/c-tabs.vue:150]:
```suggestion
      onTabClick(tab: ITab): void {
```

File: src/components/e-button.vue [hunk 10]
```diff
@@ -156,7 +154,11 @@
       },
     },
 
-    emits: ['click'],
+    emits: {
+      click(payload: Event) {
```
Comment (MB-web) [inline src/components/e-button.vue:158]:
```suggestion
      click(payload: Event): boolean {
```

File: src/stories/helpers/map-component-events-to-storybook-actions.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+/**
+ * Maps the events of the given component to an action definition for Storybooks 'argTypes'.
+ */
+export default function mapComponentEventsToStorybookActions(component: DefineComponent): IStorybookActions | null {
+  if (!component.emits || Array.isArray(component.emits)) {
+    return null;
+  }
+
+  return Object.keys(component.emits).reduce((accumulator, event) => {
```
Comment (MB-web) [inline src/stories/helpers/map-component-events-to-storybook-actions.ts:17]: Try `(accumulator: IStorybookActions ,...` and removing the cast on line 23.

**Review by MB-web — COMMENTED** [review]
Browser test: Looks good. Only minor detail: Tab link color on hover is black on black background.

In addition: maybe we should add an issue to add a nice scrolling behaviour for the tab list if it gets to long..

File: src/components/c-tabs.vue (hunk 7 above)
Comment (patric-eberle) [reply src/components/c-tabs.vue:110]: @MB-web this is JS, not TS. But sometimes we use `as PropType`. Not sure what the convention should be. Can we discuss this please?

File: src/components/c-tabs.vue (hunk 8 above)
Comment (patric-eberle) [reply src/components/c-tabs.vue:115]: @MB-web this should be defined by Vue types/interface. Can we also discuss the convention here?

**Review by MB-web — APPROVED** [review]

**Pull request merged**